test/cql-pytest/test_in_operator.py

```diff
@@ -0,0 +1,92 @@
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with Scylla.  If not, see <http://www.gnu.org/licenses/>.
+import random
+import string
+from typing import List, Any
+
+import pytest
+from cassandra import InvalidRequest
+from cassandra.cluster import Session, NoHostAvailable
+
+from util import unique_name
+
+CARTESIAN_PRODUCT_LIMIT = 100
+
+
+@pytest.fixture(scope="module")
+def table_name(cql: Session, test_keyspace: str) -> str:
+    table = f"{test_keyspace}.{unique_name()}"
+
+    cql.execute(f"CREATE TABLE {table} (p int PRIMARY KEY, a ascii, b boolean, d double,"
+                f"mai map<ascii, int>, tup frozen<tuple<text, int>>, l list<text>, s set<text>)")
+    yield table
+    cql.execute("DROP TABLE " + table)
+
+
+@pytest.fixture(scope="module")
+def rows(cql: Session, table_name: str) -> List[Any]:
+    rows = [
+        [
+            idx,
+            str(random.randint(0, 255)),
+            bool(random.randint(0, 1)),
+            random.uniform(10, 20),
+            {str(random.randint(0, 255)): random.randint(0, 1000)},
+            ("".join(random.choices(string.printable, k=random.randint(10, 20))), random.randint(0, 255)),
+            ["".join(random.choices(string.printable, k=random.randint(10, 20))) for _ in range(10)],
+            {"".join(random.choices(string.printable, k=random.randint(10, 20))) for _ in range(10)},
+        ]
+        for idx in range(200)
+    ]
+
+    insert_query = cql.prepare(
+        f"INSERT INTO {table_name} (p, a, b, d, mai, tup, l, s) VALUES ({', '.join(8 * '?')});")
+    for row in rows:
+        cql.execute(insert_query, row)
+    return rows
+
+
+@pytest.mark.parametrize("column,column_type", [
+    ("mai", "map<ascii, int>"),
+    ("l", "list<text>"),
+    ("s", "set<text>"),
+])
+def test_in_query_with_invalid_column_type(cql: Session, table_name: str, column: str, column_type: str):
+    pattern = r".*code=2200 \[Invalid query\].*IN predicates on non-primary-key columns|cannot be restricted by a" \
+              r" 'IN' relation"
+    with pytest.raises(InvalidRequest, match=pattern):
+        cql.prepare(f"SELECT * FROM {table_name} where {column} IN (?, ?) ALLOW FILTERING;")
+
+
+def test_in_query_without_allow_filtering(cql: Session, table_name: str):
+    pattern = r".*code=2200 \[Invalid query\].*IN predicates on non-primary-key columns|use ALLOW FILTERING"
+    with pytest.raises(InvalidRequest, match=pattern):
+        cql.prepare(f"SELECT * FROM {table_name} where p IN (?, ?) AND a IN (?, ?);")
+
+
+def test_in_query_with_valid_product_limit(cql: Session, rows: List[Any], table_name: str):
+    query = f"SELECT * FROM {table_name} WHERE p IN {tuple(idx for idx in range(CARTESIAN_PRODUCT_LIMIT))}" \
+            f" ALLOW FILTERING;"
+    assert len(cql.execute(query).current_rows) == CARTESIAN_PRODUCT_LIMIT
+
+
+def test_in_query_with_invalid_product_limit(cql: Session, scylla_only, table_name: str):
+    limit = 1000
```

Comment:
Again, I don't want to have tests which enshrine (i.e., "make holy, or permanent") broken behavior. I don't see in https://docs.scylladb.com/getting-started/dml/#select-statement that we ever documented the "100" limit as intentional. If we consider this as intentional, let's document it (see also https://github.com/scylladb/scylla/issues/8882 - this 100 is configurable!), and let's make sure it gives a proper error - not a NoHostAvailable (basically, an internal server error).
I think that if we don't do all of the above, it should be considered a **Scylla bug**, not a Scylla feature - Cassandra can do an IN of 200 indexes, and Scylla can't. So in that interpretation, this test should be marked xfail (a known Scylla bug), not scylla_only.

Alternatively, we can decide that you're right - this 100-limit is a real Scylla **feature** (a safe-mode feature - see #8882), so we should keep the test as scylla_only exactly like you did. You can even return it to 101 instead of 1000, it will be faster.

@avikivity you added this check and exception, what do you think?

Reply:
This is a real limit, overflowing it can OOM the coordinator.

Reply:
I don't think you are arguing that sending a IN of 101 values OOM the coordinator, but rather that in some high-enough number it will OOM, so we arbitrarily chose a limit of 100.
Is it a limit of our _implementation_ (i.e., a bug) or the API (i.e., a feature)? I.e., can't a coordinator implement the merging in a different way (perhaps we'll need to do a lot of work, but that's a different issue) so that 101 or 1000 values present no particular problems?